math/mathcore/inc/Math/CladDerivator.h

```diff
@@ -19,420 +19,221 @@
-  }
-
-  template <typename T>
-  Double_t Sin_darg0(T d) {
-    return TMath::Cos(d);
-  }
-
-  template <typename T>
-  Double_t Sin_darg0_darg0(T d) {
-    return -TMath::Sin(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0(T d) {
-    return TMath::CosH(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0_darg0(T d) {
-    return TMath::SinH(d);
-  }
-
-  template <typename T>
-  T Sq_darg0(T d) {
-    return 2 * d;
-  }
-
-  template <typename T>
-  T Sq_darg0_darg0(T d) {
-    return 2;
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0(T d) {
-    return 0.5 / TMath::Sqrt(d);
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0_darg0(T d) {
-    return 0.25 / (TMath::Sqrt(d) * d);
-  }
-
-  template <typename T>
-  Double_t Tan_darg0(T d) {
-    return 1./ TMath::Sq(TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t Tan_darg0_darg0(T d) {
-    return 2. * TMath::Sin(d) / (TMath::Cos(d) * TMath::Cos(d) * TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0(T d) {
-    return 1./ TMath::Sq(TMath::CosH(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0_darg0(T d) {
-    return -2. * TMath::SinH(d) / (TMath::CosH(d) * TMath::CosH(d) * TMath::CosH(d));
-  }
+namespace TMath {
+template <typename T>
+ValueAndPushforward<T, T> Abs_pushforward(T x, T d_x)
+{
+   return {::TMath::Abs(x), ((x < 0) ? -1 : 1) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ACos_pushforward(T x, T d_x)
+{
+   return {::TMath::ACos(x), (-1. / ::TMath::Sqrt(1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ACosH_pushforward(T x, T d_x)
+{
+   return {::TMath::ACosH(x), (1. / ::TMath::Sqrt(x * x - 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ASin_pushforward(T x, T d_x)
+{
+   return {::TMath::ASin(x), (1. / ::TMath::Sqrt(1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ASinH_pushforward(T x, T d_x)
+{
+   return {::TMath::ASinH(x), (1. / ::TMath::Sqrt(x * x + 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ATan_pushforward(T x, T d_x)
+{
+   return {::TMath::ATan(x), (1. / (x * x + 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ATanH_pushforward(T x, T d_x)
+{
+   return {::TMath::ATanH(x), (1. / (1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Cos_pushforward(T x, T d_x)
+{
+   return {::TMath::Cos(x), -::TMath::Sin(x) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> CosH_pushforward(T x, T d_x)
+{
+   return {::TMath::CosH(x), ::TMath::SinH(x) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> Erf_pushforward(T x, T d_x)
+{
+   return {::TMath::Erf(x), (2 * ::TMath::Exp(-x * x) / ::TMath::Sqrt(::TMath::Pi())) * d_x}; // Doubt
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> Erfc_pushforward(T x, T d_x)
```

Comment:
Can you please use `ValueAndPushforward<T, T>` instead of `ValueAndPushforward<Double_t, Double_t>`? This will make the custom derivatives more scalable in case ROOT ever decides to add overloads or change return types.
Please do this for all such functions.

Reply:
I replaced `ValueAndPushforward<Double_t, Double_t>` to ` ValueAndPushforward<T, T>` but then double differentiation of Log2(x) fails with following error:

```C++
/Users/vidushi/development/use-clad/builddirroot/include/Math/CladDerivator.h:119:12: error: type 'Double_t' (aka 'double') cannot be narrowed to 'int' in initializer list [-Wc++11-narrowing]
   return {::TMath::Log(x), (1. / x) * d_x};
           ^~~~~~~~~~~~~~~
<<< cling interactive line includer >>>:1:1: note: in instantiation of function template specialization 'clad::custom_derivatives::TMath::Log_pushforward<int>' requested here
^
/Users/vidushi/development/use-clad/builddirroot/include/Math/CladDerivator.h:119:12: note: insert an explicit cast to silence this issue
   return {::TMath::Log(x), (1. / x) * d_x};
           ^~~~~~~~~~~~~~~
           static_cast<int>( )
/Users/vidushi/development/use-clad/builddirroot/include/Math/CladDerivator.h:119:29: error: type 'double' cannot be narrowed to 'int' in initializer list [-Wc++11-narrowing]
   return {::TMath::Log(x), (1. / x) * d_x};
                            ^~~~~~~~~~~~~~
/Users/vidushi/development/use-clad/builddirroot/include/Math/CladDerivator.h:119:29: note: insert an explicit cast to silence this issue
   return {::TMath::Log(x), (1. / x) * d_x};
                            ^~~~~~~~~~~~~~
                            static_cast<int>( )
/Users/vidushi/development/use-clad/root_src/math/mathcore/test/CladDerivatorTests.cxx:183: Failure
```

Reply:
That's weird. I expected it to work. Can you please try to create a minimal reproducible example of this problem, and open an issue on [vgvassilev/clad](https://github.com/vgvassilev/clad).

For now, please continue to use, 'ValueAndPushforward<Double_t, Double_t>', for TMath::Log2.

Reply:
Hello Parth, I reproduced the issue again in Clad and that helped me understand the issue better. The custom derivative for Log2(x) contained `Log(2)`. `Log(2)` custom derivative was instantiated as `ValueAndPushforward<int, int> Log_pushforward<int>`. The return expression uses an initialiser list which does not allow the narrowing of expressions. Hence, producing the above errors. I resolved this by using double "2.0" instead of 2 in the definition of Log2(x) custom derivative as follows:
```TMath::Log(2.0)```